hoomd/mpcd/MPCDReversePerturbationFlow.cc

```diff
@@ -0,0 +1,339 @@
+    GPUArray<Scalar2> lo_idx(m_num_swap, m_exec_conf);
+    m_layer_lo.swap(lo_idx);
+    GPUArray<Scalar2> hi_idx(m_num_swap, m_exec_conf);
+    m_layer_hi.swap(hi_idx);
+    }
+
+mpcd::MPCDReversePerturbationFlow::~MPCDReversePerturbationFlow()
+    {
+    m_exec_conf->msg->notice(5) << "Destroying MPCDReversePerturbationFlow" << std::endl;
+    m_pdata->getBoxChangeSignal()
+        .disconnect<mpcd::MPCDReversePerturbationFlow,
+                    &mpcd::MPCDReversePerturbationFlow::requestUpdateSlabs>(this);
+    }
+
+/*!
+ * \param new_num_swap Max number of swaps
+ */
+void mpcd::MPCDReversePerturbationFlow::setNswap(unsigned int new_num_swap)
+    {
+    if (new_num_swap > m_num_swap)
+        {
+        // allocate memory if needed
+        GPUArray<Scalar2> lo_idx(new_num_swap, m_exec_conf);
+        m_layer_lo.swap(lo_idx);
+        GPUArray<Scalar2> hi_idx(new_num_swap, m_exec_conf);
+        m_layer_hi.swap(hi_idx);
+        }
+    m_num_swap = new_num_swap;
+    }
+
+//! Set the target velocity
+void mpcd::MPCDReversePerturbationFlow::setTargetMomentum(Scalar p_target)
+    {
+    m_p_target = p_target;
+    }
+
+/*!
+ * \param slab_width Slab thickness
+ */
+void mpcd::MPCDReversePerturbationFlow::setSlabWidth(Scalar slab_width)
+    {
+    // check that delta is bigger than zero
+    if (slab_width <= Scalar(0))
+        {
+        m_exec_conf->msg->error() << "MPCDReversePerturbationFlow: slab thickness " << slab_width
+                                  << " needs to be bigger than zero." << std::endl;
+        throw std::runtime_error("invalid slab thickness in MPCDReversePerturbationFlow");
+        }
+
+    m_slab_width = slab_width;
+    requestUpdateSlabs();
+    }
+
+/*!
+ * \param slab_distance Slab distance
+ */
+void mpcd::MPCDReversePerturbationFlow::setSlabDistance(Scalar slab_distance)
+    {
+    // check that delta is bigger than zero
+    if (slab_distance <= Scalar(0))
+        {
+        m_exec_conf->msg->error() << "MPCDReversePerturbationFlow: slab distance " << slab_distance
+                                  << " needs to be bigger than zero." << std::endl;
+        throw std::runtime_error("invalid slab thickness in MPCDReversePerturbationFlow");
+        }
+
+    m_slab_distance = slab_distance;
+    requestUpdateSlabs();
+    }
+
+void mpcd::MPCDReversePerturbationFlow::setSlabs()
+    {
+    const BoxDim& global_box = m_pdata->getGlobalBox();
+    const Scalar3 global_lo = global_box.getLo();
+    const Scalar3 global_hi = global_box.getHi();
+
+    const Scalar lo_pos = -m_slab_distance;
+    const Scalar hi_pos = m_slab_distance;
+    const Scalar delta_half = Scalar(0.5) * m_slab_width;
+    m_hi_pos = make_scalar2(hi_pos - delta_half, hi_pos + delta_half);
+    m_lo_pos = make_scalar2(lo_pos - delta_half, lo_pos + delta_half);
+
+    // explicit check that the two regions for velocity swapping are inside
+    // of the box entirely and don't overlap
+    if (m_hi_pos.y >= global_hi.z)
```

Comment:
This is the first spot I noticed this, but I want to make a note that we should rotate the coordinate system from the *xz* plane to the *xy* plane like I did in other geometries.